src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs

```diff
@@ -0,0 +1,973 @@
+        ///   - Remove is a command that can remove project tree items form the tree/project but not from disk. 
+        ///     For that command requests deleteOptions has DeleteOptions.None flag.
+        ///   - Delete is a command that can remove project tree items and form project and from disk. 
+        ///     For this command requests deleteOptions has DeleteOptions.DeleteFromStorage flag.
+        /// We can potentially support only Remove command here, since we don't remove Dependencies form disk, 
+        /// thus we return false when DeleteOptions.DeleteFromStorage is provided.
+        /// </summary>
+        /// <param name="nodes">The nodes that should be deleted.</param>
+        /// <param name="deleteOptions">
+        /// A value indicating whether the items should be deleted from disk as well as from the project file.
+        /// </param>
+        public override bool CanRemove(IImmutableSet<IProjectTree> nodes, 
+                                       DeleteOptions deleteOptions = DeleteOptions.None)
+        {
+            if (deleteOptions.HasFlag(DeleteOptions.DeleteFromStorage))
+            {
+                return false;
+            }
+
+            return nodes.All(node => (node.Flags.Contains(DependencyNode.GenericDependencyFlags) 
+                                        && node.BrowseObjectProperties != null)
+                                     || node.Flags.Contains(ProjectTreeFlags.Common.SharedProjectImportReference));
+        }
+
+        /// <summary>
+        /// Deletes items from the project, and optionally from disk.
+        /// Note: Delete and Remove commands are handled via IVsHierarchyDeleteHandler3, not by
+        /// IAsyncCommandGroupHandler and first asks us we CanRemove nodes. If yes then RemoveAsync is called.
+        /// We can remove only nodes that are standard and based on project items, i.e. nodes that 
+        /// are created by default IProjectDependenciesSubTreeProvider implementations and have 
+        /// DependencyNode.GenericDependencyFlags flags and IRule with Context != null, in order to obtain 
+        /// node's itemSpec. ItemSpec then used to remove a project item having same Include.
+        /// </summary>
+        /// <param name="nodes">The nodes that should be deleted.</param>
+        /// <param name="deleteOptions">A value indicating whether the items should be deleted from disk as well as 
+        /// from the project file.
+        /// </param>
+        /// <exception cref="InvalidOperationException">Thrown when <see cref="IProjectTreeProvider.CanRemove"/> 
+        /// would return <c>false</c> for this operation.</exception>
+        public override async Task RemoveAsync(IImmutableSet<IProjectTree> nodes, 
+                                               DeleteOptions deleteOptions = DeleteOptions.None)
+        {
+            if (deleteOptions.HasFlag(DeleteOptions.DeleteFromStorage))
+            {
+                throw new NotSupportedException();
```

Comment:
No, our nodes are refrences and dependencies - they should not be removed from disk. Also this is just a check and should never happen since in CanRemove we return false for DeleteFromStorage flag

---
In reply to: [74325248](https://github.com/dotnet/roslyn-project-system/pull/349#discussion_r74325248) [](ancestors = 74325248)